gensim/models/doc2vec.py

```diff
@@ -55,249 +55,560 @@
+            :attr:`gensim.utils.RULE_DISCARD`, :attr:`gensim.utils.RULE_KEEP` or :attr:`gensim.utils.RULE_DEFAULT`.
+            Note: The rule, if given, is only used to prune vocabulary during build_vocab() and is not stored as part
+            of the model.
+        callbacks : :obj: `list` of :obj: `~gensim.callbacks.Callback`
+            List of callbacks that need to be executed/run at specific stages during training.
 
         """
-        if word_vectors is None:
-            word_vectors = model.wv.syn0
-        if word_locks is None:
-            word_locks = model.syn0_lockf
-        if doctag_vectors is None:
-            doctag_vectors = model.docvecs.doctag_syn0
-        if doctag_locks is None:
-            doctag_locks = model.docvecs.doctag_syn0_lockf
-
-        word_vocabs = [model.wv.vocab[w] for w in doc_words if w in model.wv.vocab and
-                       model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
-
-        for pos, word in enumerate(word_vocabs):
-            reduced_window = model.random.randint(model.window)  # `b` in the original doc2vec code
-            start = max(0, pos - model.window + reduced_window)
-            window_pos = enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start)
-            word2_indexes = [word2.index for pos2, word2 in window_pos if pos2 != pos]
-            l1 = np_sum(word_vectors[word2_indexes], axis=0) + np_sum(doctag_vectors[doctag_indexes], axis=0)
-            count = len(word2_indexes) + len(doctag_indexes)
-            if model.cbow_mean and count > 1:
-                l1 /= count
-            neu1e = train_cbow_pair(model, word, word2_indexes, l1, alpha,
-                                    learn_vectors=False, learn_hidden=learn_hidden)
-            if not model.cbow_mean and count > 1:
-                neu1e /= count
-            if learn_doctags:
-                for i in doctag_indexes:
-                    doctag_vectors[i] += neu1e * doctag_locks[i]
-            if learn_words:
-                for i in word2_indexes:
-                    word_vectors[i] += neu1e * word_locks[i]
-
-        return len(word_vocabs)
-
-    def train_document_dm_concat(model, doc_words, doctag_indexes, alpha, work=None, neu1=None, learn_doctags=True,
-                                 learn_words=True, learn_hidden=True, word_vectors=None, word_locks=None,
-                                 doctag_vectors=None, doctag_locks=None):
-        """
-        Update distributed memory model ("PV-DM") by training on a single document, using a
-        concatenation of the context window word vectors (rather than a sum or average).
 
-        Called internally from `Doc2Vec.train()` and `Doc2Vec.infer_vector()`.
+        if 'sentences' in kwargs:
+            raise DeprecationWarning(
+                "Parameter 'sentences' was renamed to 'documents', and will be removed in 4.0.0, "
```

Comment:
Why we pass something through `*kwargs` (not as `argument=default_value`) ?

Reply:
No, this is actually good. We have changed the `sentences` argument to `documents`. It might be that the user has: `Doc2Vec(sentences=...)` in their old code, in which case the `sentences` would fall into `**kwargs` and we would just ignore it. This would result in a mysterious bug for the user. It's good that we raise this exception.